File: src/MedOps.Domain/Entities/Task.cs
namespace MedOps.Domain.Entities;

using MedOps.Domain.Enums;
using MedOps.Domain.Exceptions;

public class Task
{
    public Guid Id { get; private set; }
    public string Title { get; private set; } = string.Empty;
    public string Description { get; private set; } = string.Empty;
    public TaskStatus Status { get; private set; }
    public TaskPriority Priority { get; private set; }
    public Guid AssignedTo { get; private set; }
    public Guid CreatedBy { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime UpdatedAt { get; private set; }
    public DateOnly? DueDate { get; private set; }
    public Guid? StudyId { get; private set; }
    public ICollection<TaskAssignment> Assignments { get; private set; } = new List<TaskAssignment>();

    private Task() { }

    public Task(string title, string description, Guid assignedTo, Guid createdBy, Guid? studyId = null, DateOnly? dueDate = null)
    {
        Id = Guid.NewGuid();
        Title = title ?? throw new ArgumentNullException(nameof(title));
        Description = description ?? throw new ArgumentNullException(nameof(description));
        AssignedTo = assignedTo;
        CreatedBy = createdBy;
        StudyId = studyId;
        DueDate = dueDate;
        Status = TaskStatus.ToDo;
        Priority = TaskPriority.Medium;
        CreatedAt = DateTime.UtcNow;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Start()
    {
        if (Status != TaskStatus.ToDo && Status != TaskStatus.Cancelled)
            throw new DomainException("Only ToDo or Cancelled tasks can be started.", "INVALID_TASK_TRANSITION");
        Status = TaskStatus.InProgress;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Complete()
    {
        if (Status != TaskStatus.InProgress)
            throw new DomainException("Only InProgress tasks can be completed.", "INVALID_TASK_TRANSITION");
        Status = TaskStatus.Completed;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Cancel()
    {
        Status = TaskStatus.Cancelled;
        UpdatedAt = DateTime.UtcNow;
    }

    public void SetPriority(TaskPriority priority)
    {
        Priority = priority;
        UpdatedAt = DateTime.UtcNow;
    }

    public void AssignTo(Guid userId)
    {
        AssignedTo = userId;
        UpdatedAt = DateTime.UtcNow;
    }

    public void UpdateDetails(string title, string description)
    {
        Title = title ?? throw new ArgumentNullException(nameof(title));
        Description = description ?? throw new ArgumentNullException(nameof(description));
        UpdatedAt = DateTime.UtcNow;
    }
}

File: frontend/medops-web/src/app/services/task.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface TaskDto {
  id: string;
  title: string;
  description: string;
  status: string;
  priority: string;
  assignedTo: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateTaskDto {
  title: string;
  description: string;
  assignedTo: string;
  studyId?: string;
}

@Injectable({ providedIn: 'root' })
export class Task {
  private readonly apiUrl = '/api/tasks';

  constructor(private http: HttpClient) {}

  getAll(): Observable<TaskDto[]> {
    return this.http.get<TaskDto[]>(this.apiUrl);
  }

  getById(id: string): Observable<TaskDto> {
    return this.http.get<TaskDto>(`${this.apiUrl}/${id}`);
  }

  create(dto: CreateTaskDto): Observable<TaskDto> {
    return this.http.post<TaskDto>(this.apiUrl, dto);
  }

  start(id: string): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${id}/start`, {});
  }

  complete(id: string): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${id}/complete`, {});
  }

  cancel(id: string): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${id}/cancel`, {});
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
